### PR TITLE
Expose properties

### PR DESCRIPTION
**Pull request by zBritva**

### CODE REVIEW

File: src/app/stores/app_store.ts [hunk 1]
```diff
@@ -223,6 +224,51 @@ export class AppStore extends BaseStore {
...
+      }
+      return false;
+    }
+    const chart = this.chart;
+
+    function scaleFilter(scale: any) {
+      return !(
+        chart.elements.find((element: any) => {
+          if ((element as Specification.Object).mappings) {
+            const mappings = (element as Specification.Object).mappings;
```
Comment (stopyoukid) [inline src/app/stores/app_store.ts:248]: You could move this variable outside of the `if` statement, and do `if (mappings)`...so

```
const mappings = (element as Specification.Object).mappings;
if (mappings) {
    return hasMappedProperty(mappings, scale._id);
}
```

File: src/app/stores/app_store.ts [hunk 2]
```diff
@@ -223,6 +224,51 @@ export class AppStore extends BaseStore {
...
+            return hasMappedProperty(mappings, scale._id);
+          }
+          return false;
+        }) != null ||
+        chart.glyphs.find(glyph => {
+          return (
+            glyph.marks.find(mark => {
+              if ((mark as Specification.Object).mappings) {
+                const mappings = (mark as Specification.Object).mappings;
```
Comment (stopyoukid) [inline src/app/stores/app_store.ts:257]: Same as the above comment with this one

File: src/core/prototypes/marks/icon.ts [hunk 3]
```diff
@@ -344,4 +345,51 @@ export class IconElementClass extends EmphasizableMarkClass<
     ]);
     return widgets;
   }
+
+  public getTemplateParameters(): TemplateParameters {
+    const properties = [];
+
+    if (
+      this.object.mappings.visible &&
+      this.object.mappings.visible.type === "value"
```
Comment (stopyoukid) [inline src/core/prototypes/marks/icon.ts:354]: Not a big deal, but I don't remember if charticulator is using a newer version of typescript, but you can use optional chaining here if it does:

`if (this.object.mappings.visible?.type === "value")`

File: src/core/prototypes/marks/image.ts [hunk 4]
```diff
@@ -620,4 +621,89 @@ export class ImageElementClass extends EmphasizableMarkClass<
       { type: "y", value: cy, attribute: "cy" } as SnappingGuides.Axis
     ];
   }
+
+  public getTemplateParameters(): TemplateParameters {
+    const properties = [];
+    if (
+      this.object.mappings.strokeWidth &&
+      this.object.mappings.strokeWidth.type === "value"
```
Comment (stopyoukid) [inline src/core/prototypes/marks/image.ts:629]: Potentially optional chaining here too

File: src/core/prototypes/marks/line.ts [hunk 5]
```diff
@@ -255,4 +256,53 @@ export class LineElementClass extends EmphasizableMarkClass<
       })
     ];
   }
+
+  public getTemplateParameters(): TemplateParameters {
+    const properties = [];
+    if (this.object.mappings.visible.type !== "scale") {
```
Comment (stopyoukid) [inline src/core/prototypes/marks/line.ts:262]: Could `this.object.mappings.visible` be undefined/null here?  earlier you were checking if it was null/undefined.

File: src/app/stores/app_store.ts [hunk 6]
```diff
@@ -223,6 +224,51 @@ export class AppStore extends BaseStore {
     this.loadState(state);
   }
 
+  // removes unused scale objecs
+  private updateChartState() {
+    function hasMappedProperty(
+      mappings: Specification.Mappings,
+      scaleId: string
+    ) {
+      for (const map in mappings) {
```
[the remark this answers is not in the document]
Comment (zBritva) [reply src/app/stores/app_store.ts:233]: `mappings` is object

File: src/core/prototypes/marks/icon.ts (hunk 3 above)
Comment (zBritva) [reply src/core/prototypes/marks/icon.ts:354]: It was introduced in TS 3.7.x. Charticulator uses 3.1.6.

File: src/core/prototypes/marks/image.ts (hunk 4 above)
Comment (zBritva) [reply src/core/prototypes/marks/image.ts:629]: Fixed

File: src/core/prototypes/marks/line.ts (hunk 5 above)
Comment (zBritva) [reply src/core/prototypes/marks/line.ts:262]: Fixed

File: src/app/renderer/index.tsx [hunk 7]
```diff
@@ -67,16 +68,25 @@ export function applyColorFilter(color: Color, colorFilter: ColorFilter) {
...
+): string {
   if (!color) {
     return `rgb(0,0,0)`;
   }
-  if (colorFilter) {
-    color = applyColorFilter(color, colorFilter);
+  if (typeof color === "string") {
+    color = hexToRgb(color);
+  }
+  if (typeof color === "object") {
```
[the remark this answers is not in the document]
Comment (stopyoukid) [reply src/app/renderer/index.tsx:81]: Actually, I see what you're doing here.  The hexToRgb converts it to an object, and the next if takes that object and converts it into an rgb string

[the remark this answers is not in the document]
Comment (zBritva) [reply src/app/renderer/index.tsx:81]: No, it breaks the logic. Power BI uses HEX color format. So, color like `#FFFFFF` should be converted to `rgb(255,255,255)`

Block
```
if (typeof color === "string") {
    color = hexToRgb(color);
  }
``` 
converts HEX format of color to Color object and Color object uses to create rgb(...)

**Review by stopyoukid — APPROVED** [review]

**Comment by zBritva** [comment]
Rebased on HEAD of master

**Pull request merged**